### PR TITLE
[UNR-2129] Decouple net driver and Spatial metrics

### PR DESCRIPTION
**Pull request by ImprobableNic**
#### Description
Remove references to the Net Driver from SpatialMetrics

### CODE REVIEW

File: SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialMetrics.cpp [hunk 1]
```diff
@@ -26,11 +26,11 @@ void USpatialMetrics::Init(USpatialNetDriver* InNetDriver)
 	RPCTrackingStartTime = 0.0f;
 }
 
-void USpatialMetrics::TickMetrics()
+void USpatialMetrics::TickMetrics(float DriverTime)
```
Comment (aleximprobable) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialMetrics.cpp]: I think it's better to name the argument `NetDriverTime` or just `Time`

File: SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialMetrics.cpp [hunk 2]
```diff
@@ -84,9 +84,9 @@ void USpatialMetrics::SpatialStartRPCMetrics()
 	RPCTrackingStartTime = FPlatformTime::Seconds();
 
 	// If RPC tracking is activated on a client, send a command to the server to start tracking.
-	if (!NetDriver->IsServer())
+	if (!bIsServer && ControllerRefProvider.IsBound())
 	{
-		FUnrealObjectRef PCObjectRef = NetDriver->PackageMap->GetUnrealObjectRefFromObject(Cast<APlayerController>(NetDriver->GetSpatialOSNetConnection()->OwningActor));
+		FUnrealObjectRef PCObjectRef = ControllerRefProvider.Execute();
```
Comment (aleximprobable) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialMetrics.cpp]: We should probably check the validity of return `PCObjectRef`

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 3]
```diff
@@ -2356,3 +2357,8 @@ void USpatialNetDriver::SetSpatialDebugger(ASpatialDebugger* InSpatialDebugger)
 
 	SpatialDebugger = InSpatialDebugger;
 }
+
+FUnrealObjectRef USpatialNetDriver::GetCurrentPlayerControllerRef()
```
Comment (aleximprobable) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp]: I think it's worth checking that the `PackageMap` and the connection returned by `GetSpatialOSNetConnection` are valid.

File: SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialMetrics.cpp (hunk 2 above)
Comment (ImprobableNic) [reply SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialMetrics.cpp]: It's done indirectly below, checking the entity id for an invalid entity id

Comment (aleximprobable) [reply SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialMetrics.cpp]: Ah yes 👍

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 4]
```diff
@@ -2373,3 +2374,14 @@ void USpatialNetDriver::SetSpatialDebugger(ASpatialDebugger* InSpatialDebugger)
...
+{
+	USpatialNetConnection* NetConnection = GetSpatialOSNetConnection();
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp]: nit - we'd generally put each of these * definitions in an if statement

Comment (ImprobableNic) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp]: So nobody here has an aversion to deep-nested if statements ?
`if(USpatialNetConnection* NetConnection = GetSpatialOSNetConnection())
{
  if(APlayerController* PlayerController = Cast<APlayerController>(NetConnection->OwningActor))
  {
    if (PackageMap)
    {
      return PackageMap->GetUnrealObjectRefFromObject(PlayerController);
    }
  }
}`

**Pull request merged**